transport/gdrive-transport/src/main/java/org/apache/airavata/mft/transport/gdrive/GDriveMetadataCollector.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.airavata.mft.secret.service.GDriveSecretGetRequest;
+import org.apache.airavata.mft.secret.service.SecretServiceGrpc;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GDriveMetadataCollector implements MetadataCollector {
+
+    private String resourceServiceHost;
+    private int resourceServicePort;
+    private String secretServiceHost;
+    private int secretServicePort;
+    boolean initialized = false;
+
+    private static final Logger logger = LoggerFactory.getLogger(GDriveMetadataCollector.class);
+
+
+    @Override
+    public void init(String resourceServiceHost, int resourceServicePort, String secretServiceHost, int secretServicePort) {
+        this.resourceServiceHost = resourceServiceHost;
+        this.resourceServicePort = resourceServicePort;
+        this.secretServiceHost = secretServiceHost;
+        this.secretServicePort = secretServicePort;
+        this.initialized = true;
+    }
+
+
+    private void checkInitialized() {
+        if (!initialized) {
+            throw new IllegalStateException("GDrive Metadata Collector is not initialized");
+        }
+    }
+
+    @Override
+    public ResourceMetadata getGetResourceMetadata(String resourceId, String credentialToken) throws Exception {
+        checkInitialized();
+        ResourceServiceGrpc.ResourceServiceBlockingStub resourceClient = ResourceServiceClient.buildClient(resourceServiceHost, resourceServicePort);
+        GDriveResource gdriveResource = resourceClient.getGDriveResource(GDriveResourceGetRequest.newBuilder().setResourceId(resourceId).build());
+
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GDriveSecret gdriveSecret = secretClient.getGDriveSecret(GDriveSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        String jsonString = gdriveSecret.getCredentialsJson();
+        GoogleCredential credential = GoogleCredential.fromStream(new ByteArrayInputStream(jsonString.getBytes(StandardCharsets.UTF_8)), transport, jsonFactory);
+        if (credential.createScopedRequired()) {
+            Collection<String> scopes = DriveScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+
+        Drive drive = new Drive.Builder(transport, jsonFactory, credential).build();
+        ResourceMetadata metadata = new ResourceMetadata();
+        FileList fileList = drive.files().list()
+                .setQ("name = '"+gdriveResource.getResourcePath()+"'")
+                .setFields("files(id,name,modifiedTime,md5Checksum,size,mimeType)")
+                .execute();
+
+        for (File f : fileList.getFiles()) {
```

Comment:
Are you rewriting the metadata again and again?

Reply:
Shouldn't we expect only one entry in flieList? Do a validation before this point

Reply:
So if the user uploads a file to Google drive using MFT, the file is uploaded to a service account. If a user creates a file with same name  though Drive UI and sets the permission of file to the service account user, and when we do files.list() filtering on name of the file it considers both the files with different fileIDs and the list count would be 2. To keep both the files in sync, we need the for loop.

Reply:
There are two ways in which the file gets uploaded.
1) If the file has been created locally and uploaded via our implementation, it goes to the drive associated with the service account.
2) If the file exists in the user's drive, then it goes to the drive associated with the user's drive.

In both these cases, we use the for loop to update all files with the same name. We use this logic as the user has no way to determine the file IDs to point to a unique file.